### PR TITLE
[Intel MKL] Enabling BatchNorm test in AutoMPMkl

### PR DESCRIPTION
**Pull request by nhasabni**
This PR makes 2 fixes: one in BatchNormGrad and another one in ReLU.

BatchNormGrad fix ensures that both the inputs of BatchNormGrad are
in same layout. As per DNNL doc, lack of it would lead to sub-optimal
performance in BatchNormGrad.

ReLU fix ensures that if input of ReLU gets reordered into a blocked
layout from native layout, then output of ReLU carries correct
meta tensor.

Last change enables 1 disabled unit test in AutoMixedPrecisionMkl.

### CODE REVIEW

**Comment by penpornk** [comment]
@reedwm I was told this is related to the fix. Would you mind reviewing this?

**Comment by penpornk** [comment]
Talked to @reedwm offline. He said the mixed precision file looks good. 
I'll review the rest of the PR soon. Sorry for my delay!

**Comment by gbaned** [comment]
@nhasabni  Can you please resolve conflicts? Thanks!

**Comment by nhasabni** [comment]
@gbaned Fixed it.

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op_test.cc [hunk 1]
```diff
@@ -118,10 +124,94 @@ class CommonTestUtilities : public OpsTestBase {
...
+                                        const GraphRunnerGrad& run_mkl) {
+    int batch = 2;
+    int height = 8;
+    int width = 8;
+    int depth = 1;
+    DataType dtype = DataTypeToEnum<T>::v();
+
+    Tensor input(dtype, {batch, height, width, depth});
+    input.flat<T>() = input.flat<T>().template setRandom<random_gen_>();
+    Tensor filter(dtype, {3, 3, 1, 6});
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op_test.cc:138]: Would you mind defining variables for the sizes instead? For example,
```c++
Tensor filter(dtype, {filter_height, filter_width, in_channels, out_channels});
...
Tensor y_backprop(dtype, {batch, height, width, out_channels});
...
Tensor scale(type, {out_channels});
...
```

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op_test.cc [hunk 2]
```diff
@@ -222,6 +312,133 @@ class FusedBatchNormOpTest : public OpsTestBase {
...
+          AddInputFromArray<uint8>(conv2d_meta_output.shape(),
+                                   conv2d_meta_output.flat<uint8>());
+          AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+          AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+          AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+          AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+          TF_ASSERT_OK(RunOpKernel());
+
+          CommonTestUtilities<T> test_util;
+          test_util.PerformConversion(dtype, *GetOutput(0), *GetOutput(5),
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op_test.cc:428]: Isn't the mkl meta tensor at index 6 here?
```suggestion
          test_util.PerformConversion(dtype, *GetOutput(0), *GetOutput(6),
```
Same for `test_util_mean` and `test_util_var` (should indices 6 and 7 be 7 and 8 instead?)

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the fixes and I'm so sorry for my long delay!

**Comment by gbaned** [comment]
@nhasabni  Can you please check @penpornk's comments and keep us posted ? Thanks!

**Comment by gbaned** [comment]
@nhasabni  Any update on this PR? Please. Thanks!

File: tensorflow/core/kernels/mkl/BUILD [hunk 3]
```diff
@@ -318,6 +318,7 @@ tf_cc_test_mkl(
     srcs = ["mkl_fused_batch_norm_op_test.cc"],
     linkstatic = 1,
     deps = [
+        ":mkl_conv_op",
```
Comment (yimeisun123) [inline tensorflow/core/kernels/mkl/BUILD:321]: FYI - this change is added in the last commit but not in the original commits, since the original commits were done a while back, and the baseline changes of moving mkl kernels causes the need to adding this dependency.

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc [hunk 4]
```diff
@@ -1327,6 +1327,33 @@ class MklFusedBatchNormGradOp : public OpKernel {
...
+      MklDnnData<T> reorder_src(&cpu_engine_);
+      MklDnnData<T> reorder_diff_dst(&cpu_engine_);
+      T* diff_dst_data =
+          static_cast<T*>(const_cast<T*>(diff_dst_tensor.flat<T>().data()));
+      T* src_data =
+          static_cast<T*>(const_cast<T*>(src_tensor.flat<T>().data()));
+
+#ifdef ENABLE_MKLDNN_V1
+      // Ensure that src and diff_dst are in same blocked memory layout.
+      // As per MKL-DNN doc, this will lead to faster perf.
```
[the remark this answers is not in the document]
Comment (yimeisun123) [reply tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc]: If both are MKL tensors, they are in blocked format already and no need to adjust here for MKL-DNN.
If both are tensorflow native format, no need to adjust either as MKL-DNN can handle.
If these two inputs are in different formats, adjust the one in tensorflow native format to blocked format, as MKL-DNN has better performance for blocked format.

I have updated comments.

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op_test.cc (hunk 1 above)
Comment (yimeisun123) [reply tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op_test.cc]: Updated per suggestion in the new commit. Please check.

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op_test.cc (hunk 2 above)
Comment (yimeisun123) [reply tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op_test.cc:433]: The indices in the code are correct as they are the output indices.

**Comment by yimeisun123** [comment]
@penpornk, @gbaned - thanks for reviewing on this PR. Sorry for the long delayed response to the comments from our team. I just pushed a new commit, and also responded to the review comments in the comment section. Please let me know if any additional action needed. Thanks

File: tensorflow/core/kernels/mkl/BUILD (hunk 3 above)
Comment (penpornk) [reply tensorflow/core/kernels/mkl/BUILD:321]: Thank you for the note!

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op_test.cc (hunk 2 above)
Comment (penpornk) [reply tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op_test.cc:433]: Right. Thank you!

**Review by penpornk — APPROVED** [review]
Thank you for the fixes and explanation!
